### ISSUE TITLE
feature request: make symlinks to assets in i18n builds instead of copying

### ISSUE THREAD

**Issue by zip-fa** [angular/angular-cli#26286]
### Command

build

### Description

Hi.
Currently angular just copy assets folder to every localized build. This leads to enormous dist size (in my case every locale with assets weigths 180MB, which can be multiplied by 15 locales).

My proposal is to put assets folder in root directory, instead of copying it inside every built locale, and symlink it inside every locale. This will give us better build times, and also will fix issues where developers use absolute urls to assets, eg:
<img src=“/assets” />

### Describe alternatives you've considered

Currently, the only issue i encounter, is absolute path usage (described it in my example with img src). I make symlink to get this working:

/en/assets -> /assets

**Comment by zip-fa** [comment]
A little update:

I just took two different localized builds and compared their `hashTables` from ngsw.json. Their assets' hashes are totally identical when using application builder, so i can assume, this feature can be done without pain.

**Comment by zip-fa** [comment]
Hi guys. I see “Needs discussion” badge was added, what does it mean? Will this be done?

**Comment by alan-agius4** [comment]
`needs: discussion` means that this issue needs to be discussed further with the rest of the team. 

We already discussed this during one of our team meetings but we haven't reached a definitive conclusion on the next steps. However, we reached a consensus that if an action is taken, it's improbable that it will involve a symlink-based approach.

**Comment by dgp1130** [comment]
Spent some time discussing this in the team today. Our immediate reaction is that symlinks in the output might cause some issues for servers. In my experience, servers are typically hesitant to follow symlinks as they can often be the result of a misconfiguration issue and accidentally serve data which wasn't intended to be served. In this case, such symlinks would be internal to the served directory and (probably) safe, but I could see some servers getting this wrong or requiring explicitly enabling symlinks which server admins might not want to do.

An alternative proposal which might be better is to just emit an `assets/` directory independently of the locale. That way there is only one copy of the assets. This has the additional benefit that there is only one URL for each asset. This way if a user downloads `/assets/foo.png` and then switches locale and re-requests `/assets/foo.png`, the URL is consistent and the browser cache can respond more quickly. This would also mean absolute paths can be used for requesting assets without having to manually include the locale.

This would definitely be a breaking change for existing applications, so I don't think we can automatically make this change, it would need to be behind a flag. Not sure what the naming for this flag would be, but under i18n options we could include something like `singleAssetDir: true` to enable this behavior. I do like this approach in general, and if it works out in practice we might want to consider enabling it by default for new applications, even if we don't necessarily have a path to automatically migrating existing apps.

**Comment by zip-fa** [comment]
This sounds like a good solution. I dont think there are many people who use i18n + ssr now, so it won’t break many apps. Maybe make this approach as default and add option like { cloneAssetsForEachLocale: boolean } (which is false by default)?

**Comment by demurgos** [comment]
This feature would not only reduce build sizes considerably and improve caching, it also enables to output static files outside of the locale directories (`favicon.ico`, `robots.txt`, `.well-known`, etc.).

**Comment by spock123** [comment]
I wish I could upvote this a million times.
We have an application with a gazillion locales. For deployment we copy a local assets folder into the root folder of the dist folder, then recursively delete all the "locale" assets folders. But it's quite messy .

**Comment by martinboue** [comment]
@dgp1130, have you also considered not copying the global style files to each locale? Those defined in `projects > myapp > architect > build > options > styles`.

Is there any use case where they can be different depending on the locale? Thanks.

**Comment by dabbabi** [comment]
Hi, is there any chance to reconsider this vote, please? This is vital for new applications and especially containerised ones. I totally agree with @spock123 to upvote it more than one million time. By the way, what is the minimum requirement vote to push it back to the backlog?

**Comment by bhayward93** [comment]
I definitely think that this is something that should be considered again.

**Comment by LadaBr** [comment]
Yes current behaviour would only make sense if it was possible to change assets based on locale. Right now they are just duplicated.

**Comment by zip-fa** [comment]
Hi! Apologies for the follow-up, but I wanted to check if this feature is on the roadmap. We're running into a significant performance bottleneck where build time optimizations become irrelevant due to large marketing assets (videos/images) being duplicated across locales. With assets like 10 videos × 10MB × 10 locales, file copying takes 5+ minutes in CI while the actual build completes in under a minute. This creates a frustrating disconnect between Angular's excellent build performance and our overall deployment times. Any insights on potential solutions or timeline would be greatly appreciated!

**Comment by marcalj** [comment]
> Hi! Apologies for the follow-up, but I wanted to check if this feature is on the roadmap. We're running into a significant performance bottleneck where build time optimizations become irrelevant due to large marketing assets (videos/images) being duplicated across locales. With assets like 10 videos × 10MB × 10 locales, file copying takes 5+ minutes in CI while the actual build completes in under a minute. This creates a frustrating disconnect between Angular's excellent build performance and our overall deployment times. Any insights on potential solutions or timeline would be greatly appreciated!

You can place the assets folder in the root dist output folder. We use a bash script for that. This way you can reuse the assets for all locales. Make sure to always use the absolute path in the templates.

**Comment by spock123** [comment]
We also, as @marcalj have a build step that:
1) copies assets from /en to the root folder (to be used globally)
2) deletes assets from all the local build folders assets

However this is a bit of a hazzle to do for every build, it would be kinda nice with a "global assets" feature that did this automatically. It could be a simple "rootAssets" property in angular.json, that made sure to copy contents to the root of the dist.

**Comment by DanteDeRuwe** [comment]
Still no fix or feature for this? Seems kind of wasteful to have these assets duplicated in the bundle.

**Comment by wartab** [comment]
Another similar use-case is for applications that are served statically within Electron or NW.js.

We recently added quite a few assets to our application (mainly fonts and images) and the application size increases substantially, just because we support 4 locales.

I don't even want to imagine the nightmare that 15 locales would represent like OP. At that scale, even the JavaScript itself becomes problematic. 
I'd suggest to approach this problem differently than just sharing assets: wouldn't it be better to have an option to have a structure such as this? I'm not saying this should be the default behaviour, as I assume this would have some sort of runtime performance overhead. I also assume handling ICUs would require some extra work.

```
fr/index.html
de/index.html
en/index.html
nl/index.html
common/assets/...
common/my-component-HASH.js
```

With each `index.html` containing something like (simplified, obviously):
```javascript
ngLocale = "fr";
```

and each component, rather than hard-coding the string literal, would do something along the lines of:
```javascript
{fr:"Bonjour",de:"Hallo",en:"Hello",nl:"Hallo"}[ngLocale]
```
